# Hypothesis testing problems

## Exercise 1

You are a nutritionist researching two different types of diets to see if there is a significant difference in weight loss after one month. You choose two random groups of people; one group follows the first diet, and the other follows the second. At the end of the month, you record the weight loss (in kg) for each person:

| Diet 1 | Diet 2 |
|:-------|:-------|
| 2.0 | 3.0 |
| 2.5 | 3.2 |
| 3.0 | 3.1 |
| 2.8 | 2.9 |
| 2.3 | 2.8 |
| 2.7 | 3.0 |
| 2.5 | 3.2 |

With these data, it seeks to answer the following question: Is there a significant difference in average weight loss between people who followed the first diet and those who followed the second diet?

To draw conclusions, follow the points below:

- State the hypothesis: null and alternative hypothesis.
- Perform the test to test the hypothesis. You can use a Student's t-test.
- Analyze the conclusions.

In [2]:
import numpy as np
from scipy import stats

# Data
diet_1 = np.array([2.0, 2.5, 3.0, 2.8, 2.3, 2.7, 2.5])
diet_2 = np.array([3.0, 3.2, 3.1, 2.9, 2.8, 3.0, 3.2])

# Hypotheses
print("H0: The average weight loss is the same for both diets.")
print("H1: The average weight loss is different for both diets.\n")

# Descriptive statistics
print("Diet 1 mean:", round(diet_1.mean(), 3))
print("Diet 2 mean:", round(diet_2.mean(), 3))

# Independent samples t-test
t_stat, p_value = stats.ttest_ind(diet_1, diet_2)

print("\nT-statistic:", round(t_stat, 4))
print("P-value:", round(p_value, 6))

alpha = 0.05
if p_value < alpha:
    print("\nConclusion: Reject H0")
    print("There is a statistically significant difference in average weight loss between the two diets.")
else:
    print("\nConclusion: Fail to reject H0")
    print("There is no statistically significant difference in average weight loss between the two diets.")

H0: The average weight loss is the same for both diets.
H1: The average weight loss is different for both diets.

Diet 1 mean: 2.543
Diet 2 mean: 3.029

T-statistic: -3.5383
P-value: 0.004083

Conclusion: Reject H0
There is a statistically significant difference in average weight loss between the two diets.


### Conclusion

We used an independent samples t-test to compare the average weight loss between Diet 1 and Diet 2.

- **Null hypothesis (H0):** The average weight loss is the same for both diets.
- **Alternative hypothesis (H1):** The average weight loss is different for both diets.

The p-value is **0.004083**, which is less than **0.05**. Therefore, we reject the null hypothesis.

This means there is a statistically significant difference in average weight loss between the two diets.

## ANOVA

**ANOVA** (*Analysis of Variance*) is a statistical technique used to compare the measures of two or more groups. The idea behind ANOVA is to decompose the total variability in the data into two components: between-group variability and within-group variability:

- **Between-group variability**: This variability refers to the differences between the group means. If this variability is considerably larger than the within-group variability, it could be an indication that at least one of the group means is different.
- **Within-group variability**: This variability refers to the dispersion of the data within each group. If all groups have similar variability, then any noticeable difference in group means could be considered significant.

Hypotheses in ANOVA typically include:

- **Null hypothesis** ($H₀$): The means of all groups are equal.
- **Alternative hypothesis** ($H₁$): At least one of the group means is different.

If the ANOVA test result is significant (e.g., a p-value less than a threshold such as 0.05), this suggests that at least one group mean is different.

## Exercise 2

A farmer decides to test three different types of fertilizers to determine if one is superior in terms of corn production. The farmer plants corn on 15 identical plots and uses all three fertilizers (5 plots for each type). At the end of the season, he measures the corn yield (in kg) of each plot, with the following result:

| Fertilizer 1 | Fertilizer 2 | Fertilizer 3 |
|:-------------|:-------------|:-------------|
| 20 | 22 | 24 |
| 21 | 21 | 23 |
| 20 | 23 | 22 |
| 19 | 22 | 23 |
| 20 | 21 | 24 |

With this data, he seeks to answer the following question: Is there a significant difference in average corn yield between the three types of fertilizers?

To help you, follow the points below:

- State the hypothesis: null and alternative hypothesis.
- Perform the ANOVA test.
- Analyze the conclusions.
- If one fertilizer is better than another, how can we know it?

In [3]:
import numpy as np
from scipy import stats
import pandas as pd
from statsmodels.stats.multicomp import pairwise_tukeyhsd

# Data
fertilizer_1 = np.array([20, 21, 20, 19, 20])
fertilizer_2 = np.array([22, 21, 23, 22, 21])
fertilizer_3 = np.array([24, 23, 22, 23, 24])

# Hypotheses
print("H0: The mean corn yield is the same for all three fertilizers.")
print("H1: At least one fertilizer has a different mean corn yield.\n")

# Means
print("Fertilizer 1 mean:", round(fertilizer_1.mean(), 3))
print("Fertilizer 2 mean:", round(fertilizer_2.mean(), 3))
print("Fertilizer 3 mean:", round(fertilizer_3.mean(), 3))

# One-way ANOVA
f_stat, p_value = stats.f_oneway(fertilizer_1, fertilizer_2, fertilizer_3)

print("\nF-statistic:", round(f_stat, 4))
print("P-value:", round(p_value, 6))

alpha = 0.05
if p_value < alpha:
    print("\nConclusion: Reject H0")
    print("There is a statistically significant difference in average corn yield among the fertilizers.")
else:
    print("\nConclusion: Fail to reject H0")
    print("There is no statistically significant difference in average corn yield among the fertilizers.")

# Post-hoc test: Tukey HSD
data = pd.DataFrame({
    "yield": np.concatenate([fertilizer_1, fertilizer_2, fertilizer_3]),
    "fertilizer": (["Fertilizer 1"] * len(fertilizer_1) +
                   ["Fertilizer 2"] * len(fertilizer_2) +
                   ["Fertilizer 3"] * len(fertilizer_3))
})

print("\nTukey HSD test:")
tukey = pairwise_tukeyhsd(endog=data["yield"], groups=data["fertilizer"], alpha=0.05)
print(tukey)

H0: The mean corn yield is the same for all three fertilizers.
H1: At least one fertilizer has a different mean corn yield.

Fertilizer 1 mean: 20.0
Fertilizer 2 mean: 21.8
Fertilizer 3 mean: 23.2

F-statistic: 20.3158
P-value: 0.00014

Conclusion: Reject H0
There is a statistically significant difference in average corn yield among the fertilizers.

Tukey HSD test:
     Multiple Comparison of Means - Tukey HSD, FWER=0.05      
   group1       group2    meandiff p-adj  lower  upper  reject
--------------------------------------------------------------
Fertilizer 1 Fertilizer 2      1.8 0.0099 0.4572 3.1428   True
Fertilizer 1 Fertilizer 3      3.2 0.0001 1.8572 4.5428   True
Fertilizer 2 Fertilizer 3      1.4 0.0409 0.0572 2.7428   True
--------------------------------------------------------------


### Conclusion

We used a one-way ANOVA test to compare the mean corn yield of the three fertilizers.

- **Null hypothesis (H0):** The mean corn yield is the same for all three fertilizers.
- **Alternative hypothesis (H1):** At least one fertilizer has a different mean corn yield.

The p-value is **0.00014**, which is less than **0.05**. Therefore, we reject the null hypothesis.

This means there is a statistically significant difference in average corn yield among the fertilizers.

To see which specific fertilizers differ from each other, we also applied **Tukey HSD**. The results show that all fertilizer pairs are significantly different. Based on the sample means, **Fertilizer 3** has the highest average yield.